File: native/swift/Sources/BlocksRuntime/OIDC/OIDCAuthState.swift
import Foundation

/// Authentication state emitted by `OIDCClient.authStateChanges`.
public enum OIDCAuthState: Sendable, Equatable {
    case loading
    case signedOut
    case signedIn(OIDCUser)
}

/// The authenticated user surfaced after token exchange.
public struct OIDCUser: Sendable, Equatable, Codable {
    public let userId: String
    public let username: String
    public let groups: [String]

    public init(userId: String, username: String, groups: [String]) {
        self.userId = userId
        self.username = username
        self.groups = groups
    }

    public static func fromJSON(_ json: [String: Any]) throws -> OIDCUser {
        guard let userId = json["userId"] as? String else {
            throw OIDCError.malformedDescriptor("userId")
        }
        guard let username = json["username"] as? String else {
            throw OIDCError.malformedDescriptor("username")
        }
        let groups = (json["groups"] as? [String]) ?? []
        return OIDCUser(userId: userId, username: username, groups: groups)
    }
}

/// Configuration for a single OIDC identity provider.
public struct OIDCProviderConfig: Sendable, Equatable, Codable {
    public let authorizeUrl: String
    public let clientId: String
    public let scopes: [String]
    public let kind: String

    public init(authorizeUrl: String, clientId: String, scopes: [String], kind: String) {
        self.authorizeUrl = authorizeUrl
        self.clientId = clientId
        self.scopes = scopes
        self.kind = kind
    }

    public static func fromJSON(_ json: [String: Any]) throws -> OIDCProviderConfig {
        guard let authorizeUrl = json["authorizeUrl"] as? String else {
            throw OIDCError.malformedDescriptor("authorizeUrl")
        }
        guard let clientId = json["clientId"] as? String else {
            throw OIDCError.malformedDescriptor("clientId")
        }
        guard let scopes = json["scopes"] as? [String] else {
            throw OIDCError.malformedDescriptor("scopes")
        }
        guard let kind = json["kind"] as? String else {
            throw OIDCError.malformedDescriptor("kind")
        }
        return OIDCProviderConfig(
            authorizeUrl: authorizeUrl,
            clientId: clientId,
            scopes: scopes,
            kind: kind
        )
    }
}

File: native/dart/packages/blocks_runtime/lib/src/oidc_auth_state.dart
/// Authentication state for OIDC flows.
sealed class OidcAuthState {}

class OidcLoading extends OidcAuthState {}

class OidcSignedOut extends OidcAuthState {}

class OidcSignedIn extends OidcAuthState {
  final OidcUser user;
  OidcSignedIn(this.user);
}

/// Authenticated user info returned from token exchange.
class OidcUser {
  final String userId;
  final String username;
  final List<String> groups;

  const OidcUser({
    required this.userId,
    required this.username,
    required this.groups,
  });

  factory OidcUser.fromJson(Map<String, dynamic> json) {
    return OidcUser(
      userId: json['userId'] as String,
      username: json['username'] as String,
      // The server-relay exchange returns `{userId, username}` only (no groups),
      // so tolerate a missing/absent groups field.
      groups: (json['groups'] as List<dynamic>?)?.cast<String>() ?? const [],
    );
  }
}

/// Configuration for an OIDC identity provider.
class ProviderConfig {
  final String authorizeUrl;
  final String clientId;
  final List<String> scopes;
  final String kind;

  const ProviderConfig({
    required this.authorizeUrl,
    required this.clientId,
    required this.scopes,
    required this.kind,
  });

  factory ProviderConfig.fromJson(Map<String, dynamic> json) {
    return ProviderConfig(
      authorizeUrl: json['authorizeUrl'] as String,
      clientId: json['clientId'] as String,
      scopes: (json['scopes'] as List<dynamic>).cast<String>(),
      kind: json['kind'] as String,
    );
  }
}
